pontoon/base/models.py

```diff
@@ -2261,17 +2263,17 @@ def missing(self, locale, project=None):
         return ~Q(
             pk__in=self.get_filtered_entities(
                 locale,
-                Q(approved=True) | Q(fuzzy=True),
-                lambda x: x.approved or x.fuzzy,
+                Q(approved=True) | Q(pretranslated=True),
+                lambda x: x.approved or x.pretranslated,
                 project=project,
             )
         )
 
-    def fuzzy(self, locale, project=None):
-        """Return a filter to be used to select entities marked as "fuzzy".
+    def warnings(self, locale, project=None):
+        """Return a filter to be used to select entities with translations with warnings.
 
-        An entity is marked as "fuzzy" if all of its plural forms have a fuzzy
-        translation.
+        This filter will return an entity if at least one of its plural forms
+        has an approved, pretranslated or fuzzy translation with a warning.
```

Comment:
Why do we still care about fuzzy strings for e.g. this filter? As in, I'm struggling a bit to figure out how to conceptually fit in "fuzzy" together with "pretranslated" and "missing".

Fuzzy suggestions come from gettext's translation memory, so to me there seems to be two different possibilities for how we ought to treat them:
1. We ignore them, because Pontoon is also providing TM results, and we think they're better.
2. We take them as a sort of pretranslation.

Could we do one of the above, rather than this weird special thing? To put that another way, I think I'm arguing that we should drop this part from the [proposal](https://github.com/mozilla/pontoon/blob/master/specs/0107-pretranslated-status.md#treat-gettext-fuzzy-strings-as-missing):
> Internally, we keep using the fuzzy=True flag for Fuzzy strings, which allows us to [...]
> 2. Run quality checks on them.

Reply:
> Fuzzy suggestions come from gettext's translation memory, so to me there seems to be two different possibilities for how we ought to treat them:
> 
>     1. We ignore them, because Pontoon is also providing TM results, and we think they're better.

I'm not sure we're sure fuzzy strings are completely useless for Mozilla and other users of Pontoon. Fuzzy is a ~bug~feature of Gettext, so it should be supported.

>     2. We take them as a sort of pretranslation.

We need to differentiate between fuzzy and pretranslated in order to serialize them differently in PO files (with or without the `fuzzy` flag).

> Could we do one of the above, rather than this weird special thing?

Could you clarify more precisely which part do you find weird?

Reply:
The way I understand the term, "pretranslation" provides automated suggestions for translations from TM & MT. In some situations, we may feel that these suggestions are likely to be good enough that we could use them in production even without explicit human approval for each one.

Separately from that, "fuzzy" is how gettext encodes and communicates TM suggestions to a translator. It's effectively a subset/implementation of "pretranslated" suggestions, which allows for their representation in the `.po` file format even when unreviewed.

Therefore, if we make the automatic use-in-production of pretranslated suggestions configurable on a project+locale level, we can consider fuzzy strings as pretranslation suggestions on the input, and depending on the config, either include or not the "fuzzy" property when writing pretranslated suggestions into `.po` files.

With this approach, the whole concept of "fuzzy" goes away internally, subsumed within "pretranslated". On the other hand, if "fuzzy" remains at some level internally, that seems weird and to me communicates that we have different understandings of these terms.

Reply:
> The way I understand the term, "pretranslation" provides automated suggestions for translations from TM & MT. In some situations, we may feel that these suggestions are likely to be good enough that we could use them in production even without explicit human approval for each one.

Agreed.

> Separately from that, "fuzzy" is how gettext encodes and communicates TM suggestions to a translator. It's effectively a subset/implementation of "pretranslated" suggestions, which allows for their representation in the `.po` file format even when unreviewed.

I see concepts as similar, but I don't see "fuzzy" as a subset of "pretranslated":
1. I believe "fuzzy" strings are never perfect matches of past translations, whereas our pretranslation algorithm only takes perfect matches from the TM. 
2. "Fuzzies" are strictly avoided in production, wheres "pretranslated" are predominantly designed to be used in production.

> Therefore, if we make the automatic use-in-production of pretranslated suggestions configurable on a project+locale level, we can consider fuzzy strings as pretranslation suggestions on the input, and depending on the config, either include or not the "fuzzy" property when writing pretranslated suggestions into `.po` files.

Under this proposal, if the project is configured to store pretranslation suggestions as Unreviewed instead of Pretranslated, we need to make (non-trivial) changes to the sync process to also sync Unreviewed strings, which includes the ones suggested by humans. I find that weird. :)

> With this approach, the whole concept of "fuzzy" goes away internally, subsumed within "pretranslated". On the other hand, if "fuzzy" remains at some level internally, that seems weird and to me communicates that we have different understandings of these terms.

If we want to get rid of Fuzzy internally (because it doesn't bring much bang for the buck), I'd be more in favor of ignoring it completely than treating it as pretranslated. But I see value in landing this patch first, which makes Fuzzy almost invisible. It allows us to monitor the reactions from our users without making breaking changes internally.

Note that there's also #2017, which would help reduce the code ugliness.